mqtt/mqttd/src/app/edgehub.rs

```diff
@@ -0,0 +1,285 @@
+                        config.cert_path().to_path_buf(),
+                        config.private_key_path().to_path_buf(),
+                    )
+                })?
+        } else {
+            info!("downloading identity from edgelet");
+            download_server_certificate()
+                .await
+                .with_context(|| ServerCertificateLoadError::Edgelet)?
+        };
+
+        let broker_ready = Some(broker_ready.signal());
+        server.with_tls(tls.addr(), identity, authenticator.clone(), broker_ready)?;
+    };
+
+    Ok(server)
+}
+
+fn make_sidecars(broker_handle: &BrokerHandle, config: &Settings) -> Result<Vec<Box<dyn Sidecar>>> {
+    let mut sidecars: Vec<Box<dyn Sidecar>> = Vec::new();
+
+    let system_address = config.listener().system().addr().to_string();
+    let device_id = env::var(DEVICE_ID_ENV)?;
+
+    let settings = BridgeSettings::new()?;
+    let bridge_controller =
+        BridgeController::new(system_address.clone(), device_id.to_owned(), settings);
+    let bridge_controller_handle = bridge_controller.handle();
+
+    sidecars.push(Box::new(bridge_controller));
+
+    let mut command_handler = CommandHandler::new(system_address, &device_id);
+    command_handler.add_command(DisconnectCommand::new(&broker_handle));
+    command_handler.add_command(AuthorizedIdentitiesCommand::new(&broker_handle));
+    command_handler.add_command(PolicyUpdateCommand::new(broker_handle));
+    command_handler.add_command(BridgeUpdateCommand::new(bridge_controller_handle));
+    sidecars.push(Box::new(command_handler));
+
+    Ok(sidecars)
+}
+
+pub const WORKLOAD_URI: &str = "IOTEDGE_WORKLOADURI";
+pub const EDGE_DEVICE_HOST_NAME: &str = "EdgeDeviceHostName";
+pub const MODULE_ID: &str = "IOTEDGE_MODULEID";
+pub const MODULE_GENERATION_ID: &str = "IOTEDGE_MODULEGENERATIONID";
+
+pub const CERTIFICATE_VALIDITY_DAYS: i64 = 90;
+
+async fn download_server_certificate() -> Result<ServerCertificate> {
+    let uri = env::var(WORKLOAD_URI).context(WORKLOAD_URI)?;
+    let hostname = env::var(EDGE_DEVICE_HOST_NAME).context(EDGE_DEVICE_HOST_NAME)?;
+    let module_id = env::var(MODULE_ID).context(MODULE_ID)?;
+    let generation_id = env::var(MODULE_GENERATION_ID).context(MODULE_GENERATION_ID)?;
+    let expiration = Utc::now() + Duration::days(CERTIFICATE_VALIDITY_DAYS);
+
+    let client = edgelet_client::workload(&uri)?;
+    let cert = client
+        .create_server_cert(&module_id, &generation_id, &hostname, expiration)
+        .await?;
+
+    if cert.private_key().type_() != "key" {
+        bail!(
+            "unknown type of private key: {}",
+            cert.private_key().type_()
+        );
+    }
+
+    if let Some(private_key) = cert.private_key().bytes() {
+        let identity = ServerCertificate::from_pem_pair(cert.certificate(), private_key)?;
+        Ok(identity)
+    } else {
+        bail!("missing private key");
+    }
+}
+
+// async fn server_certificate_renewal(renew_at: DateTime<Utc>) {
```

Comment:
remove?

Reply:
forgotten piece. I need to add this one back

Reply:
added back!